kaggle_environments/core.py

```diff
@@ -538,34 +565,25 @@ def __process_specification(self, spec):
     def __agent_runner(self, agents):
         # Generate the agents.
         agents = [
-            Agent(a, self.configuration, self)
-            if a is not None
+            Agent(agent, self)
+            if agent is not None
             else None
-            for a in agents
+            for agent in agents
         ]
 
-        # Have the agents had a chance to initialize (first non-empty act).
-        initialized = [False] * len(agents)
-
         def act(none_action=None):
             if len(agents) != len(self.state):
                 raise InvalidArgument(
                     "Number of agents must match the state length")
 
-            actions = [0] * len(agents)
-            for i, agent in enumerate(agents):
-                if self.state[i]["status"] != "ACTIVE":
-                    actions[i] = None
-                elif agent is None:
-                    actions[i] = none_action
-                else:
-                    timeout = self.configuration.actTimeout
-                    if not initialized[i]:
-                        initialized[i] = True
-                        timeout += self.configuration.agentTimeout
-                    state = self.__get_shared_state(i)
-                    actions[i] = agent.act(state["observation"], timeout)
-            return actions
+            act_args = [
+                (agent, self.__get_shared_state(i), len(self.steps) == 1, self.configuration, none_action)
+                for i, agent in enumerate(agents)
+            ]
+            results = self.pool.map(act_agent, act_args)
```

Comment:
This now invokes all agents in parallel.

Reply:
Doesn't this make performance even more unpredictable than running agents in sequence, especially when all of them run on a single cpu core?
Can you elaborate a bit on how Episodes are evaluated in the online runners. According to some earlier comments, there already is some kind of process isolation between agents that is not covered here?

Reply:
Also it deadlocks for some reason when i restrict the runner to a single cpu using `taskset --cpu-list 1 kaggle-environments evaluate --environment halite --agents random random2 random3 random4`, but not when I do `taskset --cpu-list 1,2 ...`

Reply:
Also with this global state is not passed on reliably: we can't do agents anymore that use global variables for persistence.

Example:
```python
count = 0
import os
def global_agent(observation):
    global count
    count += 1
    print("player", observation["player"], "step", observation["step"], "count", (os.getpid(), count))
    return {}
```

Output: ```
player 0 step 0 count (P5, 1)
player 1 step 0 count (P6, 1)
player 2 step 0 count (P7, 1)
player 3 step 0 count (P8, 1)

player 0 step 1 count (P9, 1)
player 1 step 1 count (P0, 1)
player 2 step 1 count (P7, 2)
player 3 step 1 count (P5, 2)

player 0 step 2 count (P6, 2)
player 1 step 2 count (P8, 2)
player 2 step 2 count (P7, 3)
player 3 step 2 count (P9, 2)

player 0 step 3 count (P5, 3)
player 1 step 3 count (P0, 2)
player 2 step 3 count (P7, 4)
player 3 step 3 count (P6, 3)

player 0 step 4 count (P7, 5)
player 1 step 4 count (P9, 3)
player 2 step 4 count (P7, 6)
player 3 step 4 count (P8, 3)
```

Reply:
Also note that the example has used 6 process ids, because `processes=len(agents)`. And agents comes from `kaggle_environments.envs.halite.halite.agents` which I have modified for testing:
```python
agents = {
    "random": random_agent,
    "random2": random_agent,
    "random3": random_agent,
    "random4": random_agent,
    "global": global_agent,
    "global2": global_agent,
}
```

Reply:
I think we should separate the logging from changing execution/making the environment more efficient

Reply:
Apologies, parallelization is not quite done as I added it in last-thing on Friday. I will be disabling parallelization when running locally (and only keep parallelization in production) before releasing this PR.

Each agent has a dedicated VM with .8 vCPUs (fractional CPUs work by allocating x% of every 100 microseconds to a particular process) and 4GB of RAM. A fifth VM coordinates all of the agent VMs and performs all other trusted code updates. This change makes the fifth VM invoke the other 4 in parallel rather than round robin. The global state issue you noticed can only happen when agents are invoked inline in parallel so this shouldn't happen in production (or after I disable local parallelization).

Reply:
In general I don't think multiprocessing is a suitable form of parallelization for local execution anyway as it has a lot of idiosyncrasies around what can be parallelized.

Reply:
There's a reasonable argument for enabling parallelization locally to minimize differences with production but my perspective is that there are significant enough differences between local and production already that enabling parallelization won't provide a meaningful amount of parity. Long term I'd like to vend an all-in-one docker image that sets up the same exact VM configuration as production so users can "escalate" into a production-like environment locally to debug environment parity issues.